adapter/main.go

```diff
@@ -132,22 +135,34 @@ func (a *Adapter) makeProvider(ctx context.Context, globalHTTPTimeout time.Durat
 
 	broadcaster := record.NewBroadcaster()
 	recorder := broadcaster.NewRecorder(scheme, corev1.EventSource{Component: "keda-metrics-adapter"})
-	handler := scaling.NewScaleHandler(mgr.GetClient(), nil, scheme, globalHTTPTimeout, recorder)
+
+	kubeClientset, _ := kubernetes.NewForConfig(ctrl.GetConfigOrDie())
```

Comment:
Do we need to create a new client here? How does this impact? Could we reuse the adapter client?

Reply:
Yes, actually the adapter client `mgr.GetClient()` is managed by controller-runtime, the type is `client.Client`, which is different from the `Clientset`, and cannot be used to new informer with limited Namespace.
I have also tried to set limited Namespace for `mgr`, however it will impact all resources managed by that `mgr`, so if we want to only restrict namespace access for secret and configmap, we need to create separate informers so as not to impact other resources.

Reply:
My afraid here is that using client.Client we are using the same cache for every resource, and we have a single client for everything. Now we have potentially 2 clients, so eventually we can request the things twice, adding more load to the api-server. 
I'm not saying this is wrong, but it's a concert I have related with performance. 
I guess that we could evaluate manually the namespace we are requesting directly when we request the secrets and based on that execute the request or empty response. Using this approach, we don't need the informer and could reuse the only client we have instead of having 2.
WDYT?
CC: @kedacore/keda-contributors

Reply:
Actually I did try set the namespace when requesting, `client.Get(ctx, types.NamespacedName{Name: name, Namespace: "keda"}, secret)`
however it will still need permission to all namespaces, since the client is cache for all resources and all namespace.
I have also tried limit the namespace when `newManager`, however it will limit the namespace for all resources, since they share one client.
That's why I could only add a new `clientset`, specifically for `secret` , and later could also be used for `configmap`

Reply:
I'm still not sure about using different clients for this. I'm afraid of increasing the request against api server. Using the underlying client in the controller-runtime, we have control over the request rate, but I'm not sure with this new client.
@v-shenoy @zroubalik , WDTY?

Reply:
Unfortunately, I share the concern raised by @JorTurFer. Especially if this feature is trying to solve an edge case.

Would be beneficial to see a comparison of the network traffic before & after this change.

Reply:
Hello @zroubalik , actually the feature is not to solve an edge case, it should be a common case that normally in production env, it is not allowed for a ServiceAccount to have permission to read secrets from all namespaces as it will introduce security risk.
BTW, the feature is by default not enabled unless set RESTRICT_SECRET_ACCESS="true" in env variable. 
If not enabled, there is no performance difference compared with before; 
If enabled, I have tested with 4000 scaledobjects(each scaledobject contains 4 triggers) created in our production env, and it still works well with no performance issue.

Reply:
There are other options you can use to manage the secrets like Azure Key Vault, Hashicorp Vault or managed identities. I get your concern about the secret management, and tbh, you are totally right. The problem is how the change can impact in KEDA performance and how it interacts with the api server. KEDA could literally overload the api server if we don't control  the calls, producing a fuck up in the whole cluster.

Said this, we are working on [this feature](https://github.com/kedacore/keda/issues/2282) that potentially will allow multiple namespaced KEDA operators and transform the metrics server in a simple store for metric values, so you could just limit the secret access per namespace because you could have an operator per namespace with access to its single namespace and push the metrics to the shared metrics server, which doesn't have any access to any secret because are the operators who query the upstreams and push the values to the metrics server.

Reply:
Hello Jorge, I totally understand your concern on performance, however actually this PR is just introduce a namespaced informer for secret to only list/watch on KEDA namespace, so as no need to list/watch secrets from all namespaces, which is also a cache and will not introduce more load on APIServer, also it has proved during our performance test.

Meanwhile I have also reviewed [this feature](https://github.com/kedacore/keda/issues/2282), do you mean KEDA operators will be placed in multiple namespaces instead of only KEDA namespace? From my understanding, this feature will not help to allow KEDA to startup if removed secret from [clusterrole](https://github.com/kedacore/charts/blob/main/keda/templates/10-keda-clusterrole.yaml)  as below
```
- apiGroups:
  - ""
  resources:
  - external
  - pods
  - services
  verbs:
  - get
  - list
  - watch
```
 
Would you pls. try to remove secrets from your clusterrole as above and only add into namespaced role, then see how KEDA behaves? I thought it might help you to better understand what problem this PR is trying to solve.

Reply:
And you mentioned potentially it will allow multiple namespaced KEDA operators and transform the metrics server in a simple store for metric values, do you mean we need one KEDA operator for each application if need retrieve metrics in future?

Reply:
Hey,
No no, you don't need one operator per application, but in the near future I hope that you can have 1 operator per namespace, which basically reduces the accesses in k8s api. With the new pulling approach we are working on, the metrics server (the piece that must be global in the cluster) will be a "stupid api" without any access to anywhere, and it will pull the values from the operator/operators. 
In this scenario, is the operator who has the accesses to upstreams, so basically we could deploy operators in multiple namespaces with its own namespace as scope, limiting the accesses to its own namespace and avoiding the issue under this PR.
[This PR](https://github.com/kedacore/keda/pull/3861) does the behaviour change, being the (single) operator who gets the metrics and transforming the metrics server into the "studip" api. Once that PR is merged, we only need to add the support for multiple namespaced operators for having the namespace access isolation.

Reply:
Hey Jorge, 

To be honest, I don't think it's a good idea to have 1 operator per namespace, it will not reduce the access to k8s apiserver, instead it will increase the load to apiserver and also add huge efforts on operation and maintenance.

In our practice, we will have one namespace per application, considering the scenarios that we have 4000 applications, then do we need deploy 4000 KEDA operators in future?

Some problems might be introduced if we deployed 4000 operators:

- Each KEDA operator will create one manager which will create one client, then 4000 operators will have 4000 clients...
- Currently we only need to manage one KEDA operator and one KEDA metrics server, in future, we need manage and monitor 4000 kEDA operators in one cluster, we have 100+ clusters in our production env, which means we need actively monitor 400000 KEDA operators, that will be a nightmare to us...
- If need create / delete any application, we need also deploy / delete KEDA operator as well, this is not how a normal operator works. Normally for one operator, there will be only one instance in one cluster.
- It will not solve the secret risk since we will need manage all KEDA operators in all namespace, which means we have 4000 sa in 4000 namespaces(1 sa per namespace) to get the secret from that namespace. The only difference is previous the permission to access all secrets is granted to one sa(by clusterrole), in future the permission will be granted to 4000 sa (by namespace role)

Pls. correct me if I'm wrong.

Thanks,
Kevin

Reply:
I think the current implementation is okay.